pkg/storage/s3/checker.go

```diff
@@ -19,27 +19,31 @@ func (s *Storage) Exists(ctx context.Context, module, version string) (bool, err
 	ctx, span := observ.StartSpan(ctx, op.String())
 	defer span.End()
 
-	lsParams := &s3.ListObjectsInput{
-		Bucket: aws.String(s.bucket),
-		Prefix: aws.String(fmt.Sprintf("%s/@v", module)),
+	files := []string{"info", "mod", "zip"}
+	errChan := make(chan error, len(files))
+	defer close(errChan)
+	cancelingCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for _, file := range files {
+		go func(file string) {
+			_, err := s.s3API.HeadObjectWithContext(
+				cancelingCtx,
+				&s3.HeadObjectInput{
+					Bucket: aws.String(s.bucket),
+					Key:    aws.String(config.PackageVersionedName(module, version, file)),
+				})
+			errChan <- err
+		}(file)
 	}
-	found := make(map[string]struct{}, 3)
-	err := s.s3API.ListObjectsPagesWithContext(ctx, lsParams, func(loo *s3.ListObjectsOutput, lastPage bool) bool {
-		for _, o := range loo.Contents {
-			if _, exists := found[*o.Key]; exists {
-				log.EntryFromContext(ctx).Warnf("duplicate key in prefix %q: %q", *lsParams.Prefix, *o.Key)
-				continue
-			}
-			if *o.Key == config.PackageVersionedName(module, version, "info") ||
-				*o.Key == config.PackageVersionedName(module, version, "mod") ||
-				*o.Key == config.PackageVersionedName(module, version, "zip") {
-				found[*o.Key] = struct{}{}
+	for range files {
+		err := <-errChan
+		if err != nil {
+			var aerr awserr.Error
+			if errs.As(err, &aerr) && aerr.Code() == "NotFound" {
```

Comment:
As a side note, we should probably create a `function As` inside our `errors` package that just proxies to the standard library one, this way you don't need to import and rename to `errs` here.

Reply:
A proxy function already exists: `errors.AsErr`.